src/main/java/com/fulcrumgenomics/util/PickLongIndices.scala

```diff
@@ -0,0 +1,291 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.util
+
+import java.text.DecimalFormat
+import java.util.Random
+import java.util.regex.Pattern
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.PickLongIndices.Index
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.util.SequenceUtil
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+
+object PickLongIndices {
+  type Index = Array[Byte]
+}
+
+/** A set of indices that maintains things as both Strings and byte[]s. */
+private class IndexSet extends Iterable[Index] {
+  private val arrays = ListBuffer[Array[Byte]]()
+  private val set    = mutable.HashSet[String]()
+
+  override def iterator: Iterator[Array[Byte]] = this.arrays.iterator
+
+  /** Add an index to the set. */
+  def +=(index : Array[Byte]): Unit = {
+    this.arrays += index
+    this.set.add(new String(index))
+  }
+
+  /** Checks to see whether an index is contained in the set. */
+  def contains(index: Array[Byte]): Boolean = contains(new String(index))
+
+  /** Checks to see whether an index is contained in the set. */
+  def contains(index: String): Boolean = this.set.contains(index)
+}
+
+@clp(group=ClpGroups.Utilities, description =
+    """
+      |Picks a set of molecular indices that have at least a given number
+      |of mismatches between them.
```

Comment:
I think it makes sense to have more documentation here both for how the program works, and for the various options.

For example, how are the indices inserted into the provided adapter sequences?